xonsh/ptk_shell/shell.py

```diff
@@ -219,8 +222,8 @@ def singleline(
             # once the prompt is done, update it in background as each future is completed
             prompt_args["pre_run"] = self.prompt_formatter.start_update
 
+        events.on_pre_prompt.fire()
```

Comment:
I don't think it's ok to move `on_pre_prompt` event this far down. This way the event handler doesn't have any control over `prompt_args` anymore.

Reply:
What is "control over prompt_args"?

Reply:
The whole point of `pre_XYZ` events is to give the user control over what `XYZ` will do. For instance, the point of `pre_render` event would be to give the user the last chance to influence how the subject of the rendering is going to be rendered via event handler. It's a paradigm/pattern. This change moves the event to the point where the event handler can't influence the outcome of the respective operation (prompt rendering) at all.

Reply:
Got it. I just forget about globals that could be changed via on_pre_prompt, you're right. Then I add new event immediately before show the prompt. Could you please help with the name?
```
        events.on_NEW_EVENT_NAME.fire()
        line = self.prompter.prompt(**prompt_args)
```

Reply:
Well, the name depends on what the point of the new event is. What's the point?

Reply:
Added new `on_before_prompter_prompt` event. Changes in the name are welcome.